### ISSUE TITLE
storing a string Go value into bytea PostgreSQL column works, but reading it back fails

### ISSUE THREAD

**Issue by mitar** [jackc/pgx#1977]
**Is your feature request related to a problem? Please describe.**

With recent Go improvements [to convert strings from and to []byte without copying](https://pkg.go.dev/unsafe@master#StringData) I would like storing strings into bytea column works transparently, with as little copying as possible.

I am writing a key-value store on top of PostgreSQL and I would like to simplify storing of built-in strings so that users of my library would not have to convert them to []byte.

**Describe the solution you'd like**

It looks like storing a string Go value works already, but reading back fails with:

```
cannot scan bytea (OID 17) in binary format into *string
```

I do not see much reason why scanning bytea into string would not just work for standard types with pgx? Go allows conversion between string and []byte with simple casting, I would hope that storing into a string into the database would work the same.

**Describe alternatives you've considered**

My understanding is that I could register my own type codec, but I could not find much information about how to an existing codec: I would like that default bytea handling stays the same, only that strings are additionally scanned without copying.

**Comment by jackc** [comment]
I suspect there are unresolvable ambiguities regarding `string` <> `bytea`.

Writing the string to a PostgreSQL bytea at least appears to work because pgx sends any `string` to PostgreSQL in the text format without any conversion. pgx does the same thing when reading. Any text formatted value can be scanned directly into a `string` without any conversion.

But what happens when you read the value? In the text format, PostgreSQL will return the value hex encoded, but in binary you will get back the raw bytes.

For example: if you write the 4 bytes `jack` to a bytea you will get back the ten bytes `\x6a61636b` in the text format but the original 4 bytes in binary format. 

Even working directly in psql it can be confusing dealing with strings in `bytea`. I recommend that you instead use `text` or `varchar` if possible.

> My understanding is that I could register my own type codec, but I could not find much information about how to an existing codec: I would like that default bytea handling stays the same, only that strings are additionally scanned without copying.

If you *really* need/want to do this, you probably don't have to create a new `Codec`. I think you would only need to create your own type that implements `pgtype.BytesScanner` and `pgtype.BytesValuer`.

**Comment by mitar** [comment]
Thanks for the response.

> If you really need/want to do this, you probably don't have to create a new Codec. I think you would only need to create your own type that implements pgtype.BytesScanner and pgtype.BytesValuer.

But that would require one to scan into this custom type, not string itself.

For string itself, it seems like I would have to unregiter `TextCodec` and register a custom one? Am I mistaken? Or is there an easier way?

**Comment by mitar** [comment]
I am just thinking that since Go 1.20, there is cheap way to convert between `string` and `[]byte`:
* https://pkg.go.dev/unsafe@master#StringData
* https://pkg.go.dev/unsafe@master#SliceData

So maybe pgx could just simply convert between them as necessary.

**Comment by jackc** [comment]
> For string itself, it seems like I would have to unregister ByteaCodec and register a custom one? Am I mistaken? Or is there an easier way? Maybe just using a TryWrapScanPlanFuncs?

Yeah, either of these is a possibility, though you will still have the ambiguity of text vs. binary protocol.

> I am just thinking that since Go 1.20, there is cheap way to convert between string and []byte:

I don't really see where these could be safely used. The read buffer is reused for every row. The unsafely constructed string would be corrupted by future rows.

**Comment by vphatfla** [comment]
Little late to the party but this work for me @mitar 

```
var pw []byte
	err := DBPool.QueryRow(context.Background(), fmt.Sprintf(query, user_table_quoted), email).Scan(&pw)
	return string(pw), err
```

**Comment by mitar** [comment]
@vphatfla Yes, I would like that it works for Go string values, not Go []byte values (which already works).

**Comment by vphatfla** [comment]
@mitar, oh I see, I'm noob here, might I ask how expensive it is to repeatedly convert from []byte to string and vice versa?

**Comment by mitar** [comment]
With recent changes in Go it is very cheap. But the issue here is about ergonomics, if pgx does that internally, then it can do it a) efficiently b) without developer having to do it manually every time.

**Comment by vphatfla** [comment]
But isn't it make more sense to force developers to convert the string to []byte before saving to the postgres instead? 
My thinking is that, the data type in postgres is BYTEA, so logically, it should expect the []byte to be saved and retrieved instead of string. 
Doing it this way might help the new developers have a clear view of the system and thus, make fewer bugs in the future.

**Comment by mitar** [comment]
Sure. This is a design question for the library. It storing string into bytea an error or a feature. This is why the issue exists so that the community and maintainers can discuss the design here.

My view is that I would prefer that this is done by library automatically. Generally developers should have good understanding what is the column type and what is data type they are storing and library should try to accommodate and make this easy (especially when it is (now) possible to do it efficiently) to keep boilerplate and repeated code down.

**Comment by jackc** [comment]
> My view is that I would prefer that this is done by library automatically. Generally developers should have good understanding what is the column type and what is data type they are storing and library should try to accommodate and make this easy (especially when it is (now) possible to do it efficiently) to keep boilerplate and repeated code down.

I agree, if it can be done safely. But as I mentioned above (https://github.com/jackc/pgx/issues/1977#issuecomment-2053856959), I don't know how can be done without possibility of silently scanning the data incorrectly.

**Comment by mitar** [comment]
I reread whole thread and especially your comment https://github.com/jackc/pgx/issues/1977#issuecomment-2053856959, and I think I understand why conversions between string and []byte without copying does not really help here, because the stream is being reused.

And if I understand correctly, the issue why reading from bytea into string cannot be done inside pgx is because the protocol really lacks typing? So if I get `\x6a61636b` and want it to be read into a string, there is no way to know if this should be decoded or copied as-is. If I request that it is read into `[]byte` then it is clear that it is has to be decoded and then I can outside of pgx convert it into a string. But I thought that `RowDescription` protocol message would contain the type for the value and so if `\x6a61636b` is a bytea type, we would know that we have to decode it, before scanning it into a string? What I am missing here?

**Comment by jackc** [comment]
> And if I understand correctly, the issue why reading from bytea into string cannot be done inside pgx is because the protocol really lacks typing?

The protocol has the type information. But pgx type conversion system has a shortcut for the case of scanning a PostgreSQL value in the text format into a `string`. It simply copies the values directly. This is beneficial for several reasons.

1. It can be more correct in cases where the PostgreSQL type is not `text`/`varchar`. For example, if a PostgreSQL `timestamptz` was scanned into a Go `string` it seems more correct for the `string` to contain exactly what the server sent rather than pgx parsing the value into a `time.Time` and then formatting it as a string.
2. Performance is better when copying the value to the `string` rather than parsing and formatting the value.
3. To me, at least, it makes sense to understand scanning a text encoded value into a string as "give me exactly what PostgreSQL sent over the wire".

Unfortunately, point 3 does have surprising behavior for text-formatted `bytea` into `string`.

> But I thought that RowDescription protocol message would contain the type for the value and so if \x6a61636b is a bytea type, we would know that we have to decode it, before scanning it into a string? What I am missing here?

So yes, the information to properly decode the data is in the protocol. The problem is how to resolve the desired behavior of "give me exactly what PostgreSQL sent over the wire" with "except if is is a `bytea`".

**Comment by mitar** [comment]
> The problem is how to resolve the desired behavior of "give me exactly what PostgreSQL sent over the wire" with "except if is is a bytea".

Wouldn't it be relatively cheap to check if `RowDescription` type information says the value is `bytea` and then do something else?

**Comment by jackc** [comment]
> > The problem is how to resolve the desired behavior of "give me exactly what PostgreSQL sent over the wire" with "except if is is a bytea".
> 
> Wouldn't it be relatively cheap to check if `RowDescription` type information says the value is `bytea` and then do something else?

Yes. However, I am reluctant to special case this. Right now there is a very clear rule. No conversion happens with sending a string to PostgreSQL or scanning a text formatted value from PostgreSQL into a Go string.

This would make an exception to that. And presumably there are similar cases that one would then expect to have the same behavior such as a created type or domain with bytea underlying it.

**Comment by mitar** [comment]
I see. OK. I think one can do this with some utility function when calling into pgx, so probably it really makes sense to not include such a special case.

Thank you for your time explaining all this in this issue.
